runtime/server/controller.go

```diff
@@ -87,6 +91,47 @@ func (s *Server) ListResources(ctx context.Context, req *runtimev1.ListResources
 	return &runtimev1.ListResourcesResponse{Resources: rs}, nil
 }
 
+// WatchResourcesHandler implements an HTTP handler for runtimev1.RuntimeServiceServer
+func (s *Server) WatchResourcesHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	instanceID := r.PathValue("instance_id")
+	kind := r.URL.Query().Get("kind")
+	replay := r.URL.Query().Get("replay") == "true"
+
+	if !auth.GetClaims(ctx).CanInstance(instanceID, auth.ReadObjects) {
+		http.Error(w, "forbidden", http.StatusForbidden)
+		return
+	}
+
+	eventServer := sse.New()
+	eventServer.CreateStream("resources")
+	eventServer.Headers = map[string]string{
+		"Content-Type":  "text/event-stream",
+		"Cache-Control": "no-cache",
+		"Connection":    "keep-alive",
+	}
+
+	// Create a shim that adapts the SSE server to the WatchResources gRPC server
+	shim := &watchResourcesServerShim{
+		r:   r,
+		sse: eventServer,
+	}
+
+	// Use the existing WatchResources implementation in a goroutine
+	go func() {
+		err := s.WatchResources(&runtimev1.WatchResourcesRequest{
+			InstanceId: instanceID,
+			Kind:       kind,
+			Replay:     replay,
+		}, shim)
+		if err != nil {
+			s.logger.Error("watch resources ended with error", zap.Error(err))
+		}
```

Comment:
Don't remember if this is handled elsewhere, but you probably want to avoid error logs for `ctx` cancellations (the UI disconnects often, and we don't want ominous error logs in the console on local for that).

Related – did you do a thorough experience pass on the local experience as well (not cloud experience running on local)? E.g. `make cli` and `./rill start dev-project`?

Reply:
```
go run ./cli devtool start local --refresh-dotenv=false
go run ./cli devtool start cloud --refresh-dotenv=false
```

For the most part

Reply:
Validating this process now

Reply:
I believe this is sorted now. The currently set cloud-e2e env is set outside of ci